app/controllers/recipes_controller.rb

```diff
@@ -71,17 +70,16 @@ def set_recipe
     # Only allow a list of trusted parameters through.
     def recipe_params
       params.require(:recipe).permit(:name, :description, :preparation_time,
-                                      recipe_products_attributes: [:id, :recipe_id, :product_id, :quantity, :_destroy])
+                                      recipe_products_attributes: [:id, :recipe_id, :product_id, :quantity, :calories, :_destroy])
     end
 
-    def fetch_product_calories
-      @recipe.products.each do |product|
-        if !!(ProductCalories.exists?(product.id))
-          nutritionix_product = Nutritionix::ApiClient.new("#{product.name}")
-          ProductCalories.create(product_id: product.id,
-                              calories: nutritionix_product.calories,
-                              unit: nutritionix_product.unit,
-                              quantity: nutritionix_product.grams)
+    def calculate_recipe_product_calories
```

Comment:
I have a lot of doubts about this method here.
1. Similar to above, model is better place for such code than controller. Or even whole new service class
2. Why is it call after recipe update, as if I understood it correctly it calculates calories only for each recipe  products, not whole recipe? Or at least I can't find where `calories` field is being set for recipes.
3. Product should have relation to product_calories `has_one :product_calorie` (or `has_many :product_calories` if there can be more). And then in recipe products similar relation `has_one :product_calorie, through: :product`. After setting assotiations properlu you could just use ```rp.product_calorie```

Reply:
Now the calories are fetch after product creation. I added relation between recipe_product and product_calories through product and moved calculator to services